### PR TITLE
Enhancements for #99

### PR DESCRIPTION
**Pull request by toots**
This PR incorporates PR comments from #99 and wraps internal APIs into a private module which is re-used in the ppx syntax extension.

In the process of doing so, `dune` version had to be upgrade to `2.8` to bring in support for such private module.

This makes it possible to have internal APIs that can be shared between the main module and the ppx rewriter but not exported in the public API.

### CODE REVIEW

**Comment by bobzhang** [comment]
IMO, this is too much change for  a small enhancement, could we just rename `next_int` with `unsafe_next_int` or `internal_next_int`?

**Comment by toots** [comment]
I'm fine with that, too. Wanna do the change @bobzhang ?

**Comment by bobzhang** [comment]
@toots did the change in the original PR. I wish  we could make the changes as minimal as possible

**Comment by pmetzger** [comment]
I prefer @toots' decision here, and the change isn't actually that big. Most of it is just updates to the build infrastructure.

**Comment by toots** [comment]
Great, thanks for chiming in @pmetzger

**Comment by bobzhang** [comment]
Hi, I am really sorry but I would like to merge my commits as is. Would you mind revert my commits? I would like to give my proposed changes compared with the current proposals, sorry about that

File: src/syntax/ppx_sedlex.ml [hunk 1]
```diff
@@ -23,6 +23,17 @@ type decision_tree =
   | Table of int * int array
   | Return of int
 
+let rec simplify_decision_tree ( x : decision_tree) = 
+  match x with 
+  | Table _ | Return _ -> x 
+  | Lte (_, (Return a as l), Return b) when a = b -> l
```
Comment (hhugo) [inline src/syntax/ppx_sedlex.ml:29]: This case is redundant with the case below

**Comment by toots** [comment]
Update: this failed when deployed as `opam` package: https://github.com/ocaml/dune/issues/4839. Will revert and reconsider our options shortly.

**Comment by bobzhang** [comment]
hi @toots I am not a dune expert, but I am unclear how this solves the problem.
Your private package still needs be available to clients, otherwise it won't compiler. So instead of exporting a single unsafe function, here you are exporting a new module to clients, please correct me if I am wrong.

**Comment by bobzhang** [comment]
I am not sure if it is worthwhile to jump such loops for no good reasons, my original patch is dead simple.
Would you mind separate my commits in a single PR? 
Just double check if "using private libs" means the clients has to use dune to use sedlex?

**Comment by toots** [comment]
Hi @bobzhang. The private lib patch has been reverted, due to an issue in dune (https://github.com/ocaml/dune/issues/4839). The overall diff is pretty much you original patch: 

```diff
diff --git a/src/lib/sedlexing.ml b/src/lib/sedlexing.ml
index 865b9f7..c9fff49 100644
--- a/src/lib/sedlexing.ml
+++ b/src/lib/sedlexing.ml
@@ -159,6 +159,16 @@ let next lexbuf =
     Some ret
   end

+let __private__next_int lexbuf : int =
+  if (not lexbuf.finished) && (lexbuf.pos = lexbuf.len) then refill lexbuf;
+  if lexbuf.finished && (lexbuf.pos = lexbuf.len) then -1
+  else begin
+    let ret = lexbuf.buf.(lexbuf.pos) in
+    lexbuf.pos <- lexbuf.pos + 1;
+    if ret = (Uchar.of_int 10) then new_line lexbuf;
+    Uchar.to_int ret
+  end
+
 let mark lexbuf i =
   lexbuf.marked_pos <- lexbuf.pos;
   lexbuf.marked_bol <- lexbuf.curr_bol;
diff --git a/src/lib/sedlexing.mli b/src/lib/sedlexing.mli
index 23748ac..1af1002 100644
--- a/src/lib/sedlexing.mli
+++ b/src/lib/sedlexing.mli
@@ -151,6 +151,15 @@ val next: lexbuf -> Uchar.t option
     is exhausted, the function returns [None].
     If a ['\n'] is encountered, the tracked line number is incremented. *)

+val __private__next_int : lexbuf -> int
+(** [__private__next_int lexbuf] extracts the next code point from the
+    lexer buffer and increments to current position. If the input stream
+    is exhausted, the function returns -1.
+    If a ['\n'] is encountered, the tracked line number is incremented.
+
+    This is a private API, it should not be used by code using this module's
+    API and can be removed at any time. *)
+
 val mark: lexbuf -> int -> unit
 (** [mark lexbuf i] stores the integer [i] in the internal
     slot. The backtrack position is set to the current position. *)
diff --git a/src/syntax/ppx_sedlex.ml b/src/syntax/ppx_sedlex.ml
index afefb0b..e184313 100644
--- a/src/syntax/ppx_sedlex.ml
+++ b/src/syntax/ppx_sedlex.ml
@@ -23,6 +23,17 @@ type decision_tree =
   | Table of int * int array
   | Return of int

+let rec simplify_decision_tree ( x : decision_tree) =
+  match x with
+  | Table _ | Return _ -> x
+  | Lte (_, (Return a as l), Return b) when a = b -> l
+  | Lte (i, l, r) ->
+    let l = simplify_decision_tree l in
+    let r = simplify_decision_tree r in
+    match l, r with
+    | Return a, Return b when a = b -> l
+    | _ -> Lte (i, l,r)
+
 let decision l =
   let l = List.map (fun (a, b, i) -> (a, b, Return i)) l in
   let rec merge2 = function
@@ -156,19 +167,13 @@ let partition (name, p) =
     | Return i -> eint ~loc:default_loc i
     | Table (offset, t) ->
               let c = if offset = 0 then [%expr c] else [%expr c - [%e eint ~loc offset]] in
-        [%expr Char.code (String.get [%e evar ~loc (table_name t)] [%e c]) - 1]
+        [%expr Char.code (String.unsafe_get [%e evar ~loc (table_name t)] [%e c]) - 1]
   in
-  let body = gen_tree (decision_table p) in
+  let body = gen_tree (simplify_decision_tree (decision_table p)) in
   glb_value name
-    (pexp_function ~loc [
-      case
-        ~lhs:(ppat_construct ~loc (lident_loc ~loc "Some") (Some (pvar ~loc "uc")))
-        ~guard:None
-        ~rhs:[%expr let c = Uchar.to_int uc in [%e body]];
-      case
-        ~lhs:(ppat_construct ~loc (lident_loc ~loc "None") None)
-        ~guard:None
-        ~rhs:[%expr let c = (-1) in [%e body]]])
+    [%expr fun c ->
+      [%e body]
+    ]

 (* Code generation for the automata *)

@@ -196,7 +201,7 @@ let gen_state lexbuf auto i (trans, final) =
   let cases = Array.to_list cases in
   let body () =
     pexp_match ~loc
-      (appfun (partition_name partition) [[%expr Sedlexing.next [%e evar ~loc lexbuf]]])
+      (appfun (partition_name partition) [[%expr Sedlexing.__private__next_int [%e evar ~loc lexbuf]]])
       (cases @ [case ~lhs:[%pat? _] ~guard:None ~rhs:[%expr Sedlexing.backtrack [%e evar ~loc lexbuf]]])
   in
   let ret body = [ value_binding ~loc ~pat:(pvar ~loc (state_fun i)) ~expr:(pexp_function ~loc [case ~lhs:(pvar ~loc lexbuf) ~guard:None ~rhs:body]) ] in
```

**Comment by bobzhang** [comment]
that looks great , thanks

**Comment by hhugo** [comment]
Is there any value in keeping the second branch in `simplify_decision_tree`?

**Comment by toots** [comment]
> Is there any value in keeping the second branch in `simplify_decision_tree`?

That I cannot tell. @bobzhang ?

**Comment by bobzhang** [comment]
@hhugo It is a micro-optimization, but I don't mind if you want to get rid of it

**Comment by bobzhang** [comment]
@toots any chance to cut a new minor release for this? The size reduction is huge for us, the generated lexer code is dropped from 280K to 120K, I guess other sedlex users would benefit from this too

**Pull request merged**